frontend/ui/widget/imagewidget.lua

```diff
@@ -281,15 +288,31 @@ function ImageWidget:_render()
         self.scale_factor = self.scale_factor * DPI_SCALE
     end
 
-    -- scale to best fit container : compute scale_factor for that
+
     if self.scale_factor == 0 then
+        -- scale to best fit container: compute scale_factor for that
         if self.width and self.height then
             self.scale_factor = math.min(self.width / bb_w, self.height / bb_h)
             logger.dbg("ImageWidget: scale to fit, setting scale_factor to", self.scale_factor)
         else
             -- no width and height provided (inconsistencies from caller),
             self.scale_factor = 1 -- native image size
         end
+    elseif not self.scale_factor and self.stretch_limit_percentage then
+        -- stretch or scale to fit container, depending on self.stretch_limit_percentage
+        local screen_ratio = self.width / self.height
+        local image_ratio = bb_w / bb_h
+        local ratio_divergence_percent = math.abs(100 - image_ratio / screen_ratio * 100)
+        if ratio_divergence_percent > self.stretch_limit_percentage then
```

Comment:
From my comment above, if it is right, may be you could have this done above in a standalone if branch to avoid the duplication of the small scale bit:
```lua
if self.stretch_limit_percentage and not self.scale_factor then
   -- your stuff
   if ratio_divergence_percent > self.stretch_limit_percentage then
      self.scale_factor = 0 
   end
end

-- and the existing:
    if self.scale_factor == 0 then
        -- scale to best fit container: compute scale_factor for that
```